[WindyCall] iP

WindyCall opened the pull request:
## WindyCall: Your kind task manager
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

WindyCall frees your mind of having to remember things you need to do. It's,
- Text-based
* Command-base
+ ~~CONVENIENT~~ SUPER CONVENIENT to use

All you need to do is,
1. download it from [here](https://github.com/WindyCall/ip/releases/tag/A-Gradle)
2. drag it to the folder you want to use
3. type: **java -jar windycall-1.0.jar** in command line to get started
4. let it manage your tasks for you :smile:

One thing to notice: It is **FREE**

Features:
- [x] Managing tasks
- [x] Managing deadlines/events
- [ ] Reminders (coming soon) 

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    WindyCall chatBox = new WindyCall();
    chatBox.run();
}
```

Commit:
Add Gradle support

WindyCall committed:
Complete Level-1

WindyCall committed:
Complete Level-2

WindyCall committed:
Create inner Task class and modify previous parts accordingly

WindyCall committed:
Complete Level-3

WindyCall committed:
Create separate abstract Task file with three types of tasks. Modify addTask accordingly.

WindyCall committed:
Complete Level-4 with Inheritance used.

WindyCall committed:
Did Automated Text UI Testing using given scripts with test cases passed(test until Level-4).

WindyCall committed:
Change ChatBox name. Update input test cases.

WindyCall committed:
Create own Exception class to handle errors. Update test cases with invalid input.

WindyCall committed:
Handle errors for mark/unmark.

WindyCall committed:
Change to ArrayList to store Tasks.

WindyCall committed:
Complete Level-6 and did corresponding test.

WindyCall committed:
Complete Level-7 in branch-Level-7 branch

WindyCall committed:
Complete Level-8 in branch-Level-8

WindyCall committed:
Merge branch 'branch-Level-7'

WindyCall committed:
Resolve merging conflicts

WindyCall committed:
Create more helper classes and split codes into pieces.

WindyCall committed:
Minimally complete A-MoreOOP.
Will continue making the product more OOP in next steps.

WindyCall committed:
Organize into Pacakge

WindyCall committed:
Merge branch 'add-gradle-support' of https://github.com/WindyCall/ip

WindyCall committed:
Do JUnit Tests using Gradle

WindyCall committed:
Create a JAR File and attach it with new release

WindyCall committed:
ADD JavaDoc comments to classes and important non-private methos

WindyCall committed:
Modify codes to comply with required coding standard.

WindyCall committed:
ADD Find functionality

WindyCall committed:
Merge branch 'branch-A-JavaDoc'

WindyCall committed:
Resolve merging conflicts

WindyCall committed:
Resolve merging conflicts

yyj-02 commented on src/main/java/windycall/Deadline.java:
```diff
@@ -0,0 +1,65 @@
+    private LocalDate deadline;
+
+
+    public Deadline(String description, boolean status, String deadline) {
+        super(description, status);
+        this.deadline = processDateTime(deadline);
+    }
```
Can we move the `processDateTime` to `Parser`? In my opinion, it would be better to pass `deadline` in as `LocalDate` and let `Parser` handles the `processDateTime` as this will be more fitting with the Single Responsibility Principle.

WindyCall replied:
Nice suggestion and it makes much sense!!

yyj-02 commented on src/main/java/windycall/Deadline.java:
```diff
@@ -0,0 +1,65 @@
+        return dateTime;
+    }
+
+    private String getDeadlineFormat() {
+        return deadline.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG));
+    }
+
+    private String getDeadline() {
+        return " (by: " + getDeadlineFormat() + ")";
+    }
```
Can we change these 2 method names to make them more distinguishable while explaining the respective functionality?

WindyCall replied:
Yes, I think I haven't thought deep when naming these two methods. Thanks

yyj-02 commented on src/main/java/windycall/Event.java:
```diff
@@ -0,0 +1,65 @@
+    public String getFileFormat() {
+        return "E | " + getStatusIcon() + " | " + description + " | " + from + " | " + to + "\n";
+    }
+    private String getDeadlineFormat(LocalDate date) {
+        return date.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG));
+    }
```
There is a `getDeadlineFormat` method which I think might be confused with the methods in `Deadline` class. Can we change the method name?

yyj-02 commented on src/main/java/windycall/Parser.java:
```diff
@@ -0,0 +1,106 @@
+            System.out.println("     You should input a number to delete a task");
+            return -1;
+        }
+        int num = -1;
```
Can we use a more specific name, e.g. `index`? Similarly in other variables relating to the index.

yyj-02 commented on src/main/java/windycall/Parser.java:
```diff
@@ -0,0 +1,106 @@
+        if (command.equals("list")) return OperationType.LIST;
+        String[] parts = command.split(" ");
+        switch (parts[0]) {
+        case "mark":
+            return OperationType.MARK;
+            //break;
+        case "unmark":
+            return OperationType.UNMARK;
+            //break;
+        case "delete":
+            return OperationType.DELETE;
+            //break;
+        case "find":
+            return OperationType.FIND;
+            //break;
+        default:
+            return OperationType.ADDTASK;
+            //break;
```
You can consider using `Enum.valueOf()` method and `String.toUpper()` to achieve this in shorter lines of code. 😊

WindyCall replied:
Sure, I will have a try.

yyj-02 commented on src/main/java/windycall/Storage.java:
```diff
@@ -0,0 +1,99 @@
+    public void handleLoad(List<Task> tasks) {
+        //check if "./data" directory exist
+        File file = new File("./data");
+        File task_file = new File("./data/WindyCall.txt");;
```
Can we change this to camelCase?

yyj-02 commented on src/main/java/windycall/Storage.java:
```diff
@@ -0,0 +1,99 @@
+     *
+     * @param tasks tasks that has been changed to do users' operation on tasks
+     */
+    public void handleTaskChange(List<Task> tasks) {
```
`handleTaskChange` does not really encapsulate the functionality of this method. Should we use `writeTasksToFile` instead?

yyj-02 commented on src/main/java/windycall/Ui.java:
```diff
@@ -0,0 +1,64 @@
+
+    }
+
+    public static void space() {
+        System.out.print("     ");
+    }
+
+    public static void line() {
+        space();
+        System.out.println("--------------------------------------------------");
+    }
```
For method names, it is best to use an action verb. Can we change it to `printSpace` and `printLine` respectively?

yyj-02 commented on src/main/java/windycall/Ui.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public void greeting() {
```
Similarly for this

yyj-02 commented on src/main/java/windycall/Ui.java:
```diff
@@ -0,0 +1,64 @@
+        line();
+    }
+
+    public void byeWords() {
```
And this too

yyj-02 commented on src/test/java/windycall/DeadlineTest.java:
```diff
@@ -0,0 +1,25 @@
+
+public class DeadlineTest {
+    @Test
+    public void standardDateFormatTest() {
```
I think the usual convention for testing method names should start with `should...` to say its expected behaviour, i.e. `shouldParseFormattedDate`. Do you want to follow that convention?

PeanutButters93 replied:
I think the method name is fine.
According to the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html), the test function can be featureUnderTest_testScenario_expectedBehavior(), where the testScenario and expectedBehaviour are optional.

yyj-02 commented on src/main/java/windycall/Task.java:
```diff
@@ -0,0 +1,52 @@
+
+    public Task(String description, boolean status) {
+        this.description = description;
+        this.isDone = status;
```
Can we change `status` to `isDone` also?

WindyCall replied:
I've looked up all your suggestions and they are really helpful for me. Thanks for your kind and careful reminder!!!

Btw, have we met before, why I felt like you are quite familiar hahah?

yyj-02 reviewed (commented):
Well done! 👍👍 I think your variable namings are generally very well thought out and self-explanatory. If I have to nitpick, however, I would suggest that you can rename the testing method names to start with `should...`, stating the expected behaviour. But overall very good effort! 🔥

WindyCall committed:
ADD OperationHandler and related Handler.
Make it more OOP

WindyCall committed:
COMPLETE AddTaskHandler

WindyCall committed:
FIX some bugs and REFINE UI and Parser.

PeanutButters93 commented on src/main/java/windycall/Ui.java:
```diff
@@ -0,0 +1,81 @@
+
+    }
+
+    public static void space() {
+        System.out.print("     ");
+    }
```
I notice a lot of ui.space() followed by System.out.println().
Perhaps consider abstracting further and have a ui.println() ?
It would also make it more clear that space is writing something to the user.

PeanutButters93 commented on src/main/java/windycall/WindyCall.java:
```diff
@@ -0,0 +1,76 @@
+
+      private List<Task> tasks;
+      private Storage storage;
+      private Ui ui;
```
Having the Ui object be called ui and frequently calling both could be confusing. Might I suggest renaming the ui to uiObject? If you choose to do so, there is a feature in Intelij to automatically edit all instances of a phrase (shift+F6).

WindyCall replied:
Sure, it makes sense. Thanks for that.

PeanutButters93 commented on src/main/java/windycall/OperationHandler.java:
```diff
@@ -0,0 +1,7 @@
+
+public abstract class OperationHandler {
+
+    public abstract void handle(String userCommand);
```
I see there are static handle functions and an instance handle functions. Maybe consider renaming or removing one? I don't believe this (static) is used. I believe you may have tried to override this static function. I also tried that and realised it was not possible.

WindyCall replied:
Yes, I do have this issue when coding. So far I just use static to make it functionable, and I think no need to do the renaming because I can overload this handle methods specifically in different classes while maintaining ```public void handle(String)```.
Anyways, thanks for sharing your own experience, I'll have a try!!

PeanutButters93 reviewed (commented):
Looks very good.
Some minor changes suggested.

WindyCall committed:
FINISH Checkstyle

WindyCall committed:
ADD models for GUI without connection with previous product

WindyCall committed:
COMPLETE adding GUI to WindyCall.

WindyCall committed:
Merge branch 'branch-Level-10'

WindyCall committed:
Add some assertions to index related methods.

Index related methods handle wrong index and out-of bound index using
exceptions.

Important assumptions in this part are not shown explicitly.

Let's add some assertions that explicitly shows these assumptions so
that it improves readability.

Using assertions in this case is preferable as it does convey assumption

WindyCall committed:
Refactor classes and Create more sub-packages.

The application only has one main package and does not provide a good
abstraction on functionality of different classes.

Let's add some sub-packages to arrange classes according to their
functionality.

Now functionality of different classes are clearly shown.

WindyCall committed:
Merge pull request #1 from WindyCall/branch-A-Assertions

Add some assertions to index related methods.

WindyCall committed:
Merge branch 'master' into branch-A-CodeQuality

WindyCall committed:
Merge branch 'branch-A-CodeQuality'

WindyCall committed:
Complete C-Tagging

WindyCall committed:
Merge branch 'branch-C-Tagging'

WindyCall committed:
Fix some bugs in loading file.

WindyCall committed:
Handle more potential index-related errors.

WindyCall committed:
Handle errors related to time format.

WindyCall committed:
Make OperationHandler related classes more OOP

WindyCall committed:
Make Gui prettier.

WindyCall committed:
Upload representative screenshot.

WindyCall committed:
Complete final code standard check.

WindyCall committed:
Improve code quality.

WindyCall committed:
Reupload Ui.png and refactor images.

WindyCall committed:
Complete User Guide.

WindyCall committed:
Add greetings from WindyCall.

WindyCall committed:
Fix Java version warning.
